include/pybind11/pytypes.h

```diff
@@ -346,6 +346,30 @@ class error_already_set : public std::runtime_error {
     /// the given tuple.
     bool matches(handle exc) const { return PyErr_GivenExceptionMatches(m_type.ptr(), exc.ptr()); }
 
+    // Raise a new exception using this exception as its cause
+    void caused(PyObject *type, const char *message) {
+        // from cpython/errors.c
+        PyObject *exc, *val, *val2, *tb; 
+        exc = m_type.ptr();
+        val = m_value.ptr();
+        tb = m_trace.ptr();
+        
+        PyErr_NormalizeException(&exc, &val, &tb);
+        if (tb != nullptr) {
+            PyException_SetTraceback(val, tb);
+            m_trace.release();
+        }
+        m_type.release();
+
+        PyErr_SetString(type, message);
+        PyErr_Fetch(&exc, &val2, &tb);
+        PyErr_NormalizeException(&exc, &val2, &tb);
+        Py_INCREF(val);
+        PyException_SetCause(val2, val);
+        PyException_SetContext(val2, val);
+        PyErr_Restore(exc, val2, tb);
```

Comment:
I think a better implementation of this would actually hold onto the newly created exception. The caller can call `error_already_set.restore()` if that's what they want.